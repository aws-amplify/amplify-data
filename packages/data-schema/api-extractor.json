{
  "extends": "../../api-extractor.base.json",
  "mainEntryPointFilePath": "<projectFolder>/dist/esm/src/index.d.ts"
}
