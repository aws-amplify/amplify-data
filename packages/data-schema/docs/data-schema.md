<!-- Do not edit this file. It is automatically generated by API Documenter. -->

[Home](./index.md) &gt; [@aws-amplify/data-schema](./data-schema.md)

## data-schema package

## Enumerations

<table><thead><tr><th>

Enumeration


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[ModelRelationshipTypes](./data-schema.modelrelationshiptypes.md)


</td><td>

Model relationship types


</td></tr>
</tbody></table>

## Interfaces

<table><thead><tr><th>

Interface


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[EnumType](./data-schema.enumtype.md)


</td><td>

Enum type definition content


</td></tr>
</tbody></table>

## Namespaces

<table><thead><tr><th>

Namespace


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[a](./data-schema.a.md)


</td><td>


</td></tr>
</tbody></table>

## Type Aliases

<table><thead><tr><th>

Type Alias


</th><th>

Description


</th></tr></thead>
<tbody><tr><td>

[Authorization](./data-schema.authorization.md)


</td><td>

Container for authorization schema definition content.


</td></tr>
<tr><td>

[ClientSchema](./data-schema.clientschema.md)


</td><td>


</td></tr>
<tr><td>

[CombinedModelSchema](./data-schema.combinedmodelschema.md)


</td><td>

CombinedModel schema definition interface


</td></tr>
<tr><td>

[CustomOperation](./data-schema.customoperation.md)


</td><td>

Custom operation definition interface


</td></tr>
<tr><td>

[CustomType](./data-schema.customtype.md)


</td><td>

Custom type container


</td></tr>
<tr><td>

[Json](./data-schema.json.md)


</td><td>

A precise, recursive Json type blows the type calculation stack without installing explicit `Json extends T ? short-circuit : ...` type checks all over the place. We may take that on later. But, this is a good-enough approximation for now.


</td></tr>
<tr><td>

[ModelDefaultIdentifier](./data-schema.modeldefaultidentifier.md)


</td><td>

Model default identifier


</td></tr>
<tr><td>

[ModelField](./data-schema.modelfield.md)


</td><td>

Public API for the chainable builder methods exposed by Model Field. The type is narrowing e.g., after calling .array() it will be omitted from intellisense suggestions


</td></tr>
<tr><td>

[ModelRelationshipField](./data-schema.modelrelationshipfield.md)


</td><td>

Model relationship field definition interface


</td></tr>
<tr><td>

[ModelRelationshipTypeArgFactory](./data-schema.modelrelationshiptypeargfactory.md)


</td><td>

Model relationship type definition content


</td></tr>
<tr><td>

[ModelSchema](./data-schema.modelschema.md)


</td><td>

Model schema definition interface


</td></tr>
<tr><td>

[ModelType](./data-schema.modeltype.md)


</td><td>

Model type definition interface


</td></tr>
<tr><td>

[Nullable](./data-schema.nullable.md)


</td><td>

Field type arg mutators


</td></tr>
<tr><td>

[RDSModelSchema](./data-schema.rdsmodelschema.md)


</td><td>

RDSModel schema definition interface


</td></tr>
<tr><td>

[RefType](./data-schema.reftype.md)


</td><td>

Reference type definition interface


</td></tr>
<tr><td>

[ResourceAuthorization](./data-schema.resourceauthorization.md)


</td><td>

Container for representing the authorization of function resources


</td></tr>
</tbody></table>
