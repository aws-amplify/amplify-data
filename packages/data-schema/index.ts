import * as a from './src';
import { ClientSchema } from './src/ClientSchema';

export { a };

export type { ClientSchema };

export * from './src/runtime/client';
export {
  addSchemaToClient,
  addSchemaToClientWithInstance,
} from './src/runtime';
