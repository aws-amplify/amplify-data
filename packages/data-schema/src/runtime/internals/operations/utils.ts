// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
// import { GraphQLFormattedError } from '@aws-amplify/data-schema-types';

/**
 * Handle errors for list return types (list and index query operations)
 */
export function handleListGraphQlError(error: any) {
  if (error?.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: [],
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 * Handle errors for singular return types (create, get, update, delete operations)
 */
export function handleSingularGraphQlError(error: any) {
  if (error.errors) {
    // graphql errors pass through
    return {
      ...error,
      data: null,
    } as any;
  } else {
    // non-graphql errors are re-thrown
    throw error;
  }
}

/**
 *
 * @param resolver
 * @returns
 */
export function selfAwareAsync<T>(
  resolver: (promise: Promise<T>) => Promise<T>,
): Promise<T> {
  let resolve: (value: T) => void;
  let reject: (reason?: any) => void;

  const resultPromise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  });

  resolver(resultPromise)
    .then((result) => {
      resolve(result);
    })
    .catch((error) => {
      reject(error);
    });

  return resultPromise;
}
