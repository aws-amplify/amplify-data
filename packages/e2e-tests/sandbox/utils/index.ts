import { deploySandbox, deploySandboxWatch } from './deploy';
import { teardownSandbox } from './teardown';
import { updateSchema } from './schemaUtils';

/**
 * Sandbox operations can sometimes be time consuming, so we increase the Jest
 * setup / teardown hook timeouts to 10 minutes.
 */
export const sandboxTimeout: number = 600000;

export const outputsFileName = 'amplify_outputs.json';

export { deploySandbox, deploySandboxWatch };
export { teardownSandbox };
export { updateSchema };
