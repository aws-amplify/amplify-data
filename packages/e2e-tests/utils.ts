import { Amplify } from 'aws-amplify';
import { generateClient } from 'aws-amplify/data';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';

export const statuses = {
  success: {
    icon: '✅',
    label: 'Success',
  },
  fail: {
    icon: '❌',
    label: 'Fail',
  },
} as const;

const _testCase = (success: boolean): any =>
  success ? statuses.success : statuses.fail;

// TODO: use imported type from `aws-amplify/data` once it's fixed
type Client = ReturnType<typeof configureAmplifyAndGenerateClient>;

export type TestCase = {
  label: string;
  action: (client: Client) => Promise<boolean>;
};

/**
 * Given an array of test cases, runs each test case
 * and asserts that the result is successful
 * @param testCases
 * @param client
 */
export async function runTestCases(testCases: TestCase[]) {
  const client = (global as any).client as Client;

  for (const testCase of testCases) {
    test(`${testCase.label}`, async () => {
      let result: any;

      try {
        const response = await testCase.action(client);
        result = _testCase(response);
      } catch (error) {
        result = _testCase(false);
      }

      console.log(`${result.icon} ${testCase.label}`);
      expect(result.label).toBe(statuses.success.label);
    });
  }
}

/**
 * Configures Amplify and returns API client
 * @returns API client
 */
export function configureAmplifyAndGenerateClient() {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs);

  const client = generateClient<Schema>();

  return client;
}
