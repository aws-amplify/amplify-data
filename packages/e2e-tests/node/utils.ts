import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
// TODO: the dep exists, investigate why it's not being resolved:
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  disableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param disableDebugLogging - disables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  disableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export async function pause(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

type EstablishWebsocketParams = {
  disableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const establishWebsocket = ({
  disableConnectionStateLogging = false,
}: EstablishWebsocketParams): any => {
  (global as any).WebSocket = WebSocket;

  /**
   * Debug logs are enabled by default to assist with CI failure debugging.
   * Option to disable is included for local development, in the event that
   * the debug logs are too verbose.
   */
  if (!disableConnectionStateLogging) {
    // https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/#subscription-connection-status-updates
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

export const handleErrorsAndData = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};
