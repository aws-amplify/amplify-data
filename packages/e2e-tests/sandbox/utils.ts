import { Amplify } from 'aws-amplify';
import {
  CONNECTION_STATE_CHANGE,
  ConnectionState,
  generateClient,
} from 'aws-amplify/data';
import { Hub, ConsoleLogger } from 'aws-amplify/utils';
import type { Schema } from './amplify/data/resource';
import outputs from './amplify_outputs.json';
import { WebSocket } from 'ws';

// TODO: use imported type from `aws-amplify/data` once it's fixed
export type Client = ReturnType<typeof generateClient<Schema>>;

type ConfigureAmplifyAndGenerateClientParams = {
  enableDebugLogging?: boolean;
  amplifyOptions?: any;
  apiClientOptions?: any;
};

/**
 * Configures Amplify and returns API client
 * @param enableDebugLogging - enables debug logging
 * @param amplifyOptions - options to pass to Amplify.configure
 * @param apiClientOptions - options to pass to generateClient
 * @returns API client
 */
export const configureAmplifyAndGenerateClient = ({
  enableDebugLogging = false,
  amplifyOptions = {},
  apiClientOptions = {},
}: ConfigureAmplifyAndGenerateClientParams): Client => {
  console.log('configuring Amplify and generating client..');
  Amplify.configure(outputs, amplifyOptions);

  if (enableDebugLogging) {
    ConsoleLogger.LOG_LEVEL = 'DEBUG';

    Hub.listen('core', (data: any) => {
      if (data.payload.event === 'configure') {
        console.log('API configuration details:', data.payload.data.API);
      }
    });
  }

  return generateClient<Schema>(apiClientOptions);
};

type AddWebsocketPolyfillParams = {
  enableConnectionStateLogging?: boolean;
};

/**
 * Adds the WebSocket API globally.
 * Subscriptions do not work in Node.js environment without the WebSocket API.
 */
export const addWebsocketPolyfill = ({
  enableConnectionStateLogging = false,
}: AddWebsocketPolyfillParams) => {
  (global as any).WebSocket = WebSocket;

  if (enableConnectionStateLogging) {
    Hub.listen('api', (data: any) => {
      const { payload } = data;
      if (payload.event === CONNECTION_STATE_CHANGE) {
        const connectionState = payload.data.connectionState as ConnectionState;
        console.log(
          'Connection state has changed. Connection state: ',
          connectionState,
        );
      }
    });
  }
};

/**
 * Util that takes model operation response and throws an error if errors
 * are present, or if the data is undefined.
 * @param response - model operation response
 * @param operation - operation name
 * @returns data from response
 */
export const expectDataReturnWithoutErrors = <T>(
  response: { data?: T; errors?: any },
  operation: string,
): T => {
  if (response.errors) {
    console.log(`error on ${operation}:`, response.errors);
    throw new Error(JSON.stringify(response.errors));
  }

  if (!response.data) {
    throw new Error(`no response data for ${operation}`);
  }

  return response.data;
};

/**
 * Function that listens for "Subscription ack" event, and once it receives it,
 * resolves.
 * Jest will timeout if the test does not complete within the default timeout of
 * 5 seconds, so no timeout handling is required.
 */
export const waitForSubscriptionAck = () =>
  new Promise<void>((resolve) => {
    const cancel = Hub.listen('api', async (data: any) => {
      const { payload } = data;
      // TODO: update once we add export:
      if (payload.event === 'Subscription ack') {
        console.log('Subscription ack received:', payload.data);
        // Stop listening for msgs once we receive the ack:
        cancel();
        // Resolve so test execution can continue:
        resolve();
      }
    });
  });
