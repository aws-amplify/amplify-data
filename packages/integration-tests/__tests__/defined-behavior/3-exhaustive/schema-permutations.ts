import { a, ClientSchema } from '@aws-amplify/data-schema';
import { buildAmplifyConfig } from '../../utils';

const fieldTypes = [
  'boolean',
  'date',
  'datetime',
  'email',
  'float',
  'id',
  'integer',
  'ipAddress',
  'json',
  'phone',
  'string',
  'time',
  'timestamp',
  'url',
] as const;

// non-primitive field types
// enum
// customType

describe('primitive field types', () => {
  describe('generate expected graphql', () => {
    for (const fieldType of fieldTypes) {
      test(fieldType, async () => {
        const schema = a
          .schema({
            Model: a.model({
              optionalField: a[fieldType](),
              requiredField: a[fieldType]().required(),
            }),
          })
          .authorization([a.allow.public()]);
        const graphql = schema.transform().schema;
        expect(graphql).toMatchSnapshot();
      });
    }
  });
});

// how can we do this better to ensure we've got full coverage over auth graphql generation?
describe('auth rules', () => {
  for (const [name, rule] of Object.entries({
    'owner()': a.allow.owner(),
    'owner().inField()': a.allow.owner().inField('field'),
    'owner().inField().to()': a.allow
      .owner()
      .inField('field')
      .to(['create', 'delete']),
    'owner(userpool).inField().to().': a.allow
      .owner('userPools')
      .to(['create', 'list']),
  })) {
    test(name, () => {
      const schema = a.schema({
        Model: a
          .model({
            field: a.string(),
          })
          .authorization([rule]),
      });
      const graphql = schema.transform().schema;
      expect(graphql).toMatchSnapshot();
    });
  }
});
