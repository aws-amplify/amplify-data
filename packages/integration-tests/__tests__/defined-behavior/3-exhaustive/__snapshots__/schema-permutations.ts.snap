// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth rules owner() 1`] = `
"type Model @model @auth(rules: [{allow: owner, ownerField: "owner"}])
{
  id: ID! @primaryKey
  field: String
  owner: String
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`auth rules owner().inField() 1`] = `
"type Model @model @auth(rules: [{allow: owner, ownerField: "field"}])
{
  id: ID! @primaryKey
  field: String
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`auth rules owner().inField().to() 1`] = `
"type Model @model @auth(rules: [{allow: owner, operations: [create, delete], ownerField: "field"}])
{
  id: ID! @primaryKey
  field: String
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`auth rules owner(userpool).inField().to(). 1`] = `
"type Model @model @auth(rules: [{allow: owner, provider: userPools, operations: [create, list], ownerField: "owner"}])
{
  id: ID! @primaryKey
  field: String
  owner: String
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql boolean 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: Boolean
  requiredField: Boolean!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql date 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSDate
  requiredField: AWSDate!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql datetime 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSDateTime
  requiredField: AWSDateTime!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql email 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSEmail
  requiredField: AWSEmail!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql float 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: Float
  requiredField: Float!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql id 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: ID
  requiredField: ID!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql integer 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: Int
  requiredField: Int!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql ipAddress 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSIPAddress
  requiredField: AWSIPAddress!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql json 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSJSON
  requiredField: AWSJSON!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql phone 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSPhone
  requiredField: AWSPhone!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql string 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: String
  requiredField: String!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql time 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSTime
  requiredField: AWSTime!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql timestamp 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSTimestamp
  requiredField: AWSTimestamp!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;

exports[`primitive field types generate expected graphql url 1`] = `
"type Model @model @auth(rules: [{allow: public}])
{
  id: ID! @primaryKey
  optionalField: AWSURL
  requiredField: AWSURL!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}"
`;
